File: core0/us/sdk/wallet/cs/us/gov/socket/peer_t.cs
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the AGPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
using System.Net.Sockets;

namespace us.gov.socket
{
    public class peer_t : client
    {
        public peer_t() : base()
        {
        }

        public peer_t(Socket sock): base(sock)
        {            
        }

        public bool is_slow()
        {
            return false;
        }

        public bool ping()
        {
            return send3(new datagram((ushort)protocol.gov_socket_ping, "ping")) == null;
        }

        public void process_pong()
        {
        }

        public virtual bool process_work(datagram d)
        {
            if (d.service.Equals(protocol.gov_socket_ping))
            {
                send3(new datagram((ushort)protocol.gov_socket_pong, "pong"));
            }
            else if (d.service.Equals(protocol.gov_socket_pong))
            {
                process_pong();                
            }
            else {
                return false;
            }
            return true;
        }

    }
}

File: core0/us/gov/relay/peer_t.cpp
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the GPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
#include "api.h"
#include <us/gov/io/seriable.h>
#include <us/gov/socket/daemon0_t.h>

#include "peer_t.h"

#define loglevel "gov/relay"
#define logclass "peer_t"
#include "logs.inc"
#include <us/gov/socket/dto.inc>

using namespace us::gov::relay;
using c = us::gov::relay::peer_t;


bool peer_t::process_work(datagram* d) {
    assert(d != nullptr);
    #ifdef has_us_gov_relay_api
    using namespace us::gov::protocol;
    switch(d->service) {
        #include <us/api/generated/gov/c++/relay/hdlr_svc-router>
    }
    #endif
    return b::process_work(d);
}

#ifdef has_us_gov_relay_api

#include <us/api/generated/gov/c++/relay/hdlr_svc_handler-impl>


//------------------apitool - API Spec defined @ us/api/generated/gov/c++/relay/hdlr_local-impl

ko c::handle_push(push_in_dst_t&& o_in) {
    log("push");
    /// in:
    ///     hash_t tid;
    ///     uint16_t code;
    ///     blob_t blob;


    // Implement here.

    return KO_10210;
}

//-/----------------apitool - End of API implementation.


#include <us/api/generated/gov/c++/relay/cllr_rpc-impl>

#endif

